src/file-write.ts

```diff
@@ -0,0 +1,135 @@
+import {
+    fsReadPromise,
+    fsRenamePromise,
+    fsUnlinkPromise,
+    fsWritePromise,
+    getNextBufferSubarrayAsync,
+    getNextBufferSubarraySync,
+    processFile,
+    processFileAsync
+} from "./util-file"
+import * as tmp from 'tmp'
+import * as path from 'path'
+import * as fs from 'fs'
+import { findId3TagPosition, getId3TagSize } from "./id3-tag"
+
+const FileBufferSize = 20 * 1024 * 1024
+
+export function writeId3TagToFileSync(filepath: string, id3Tag: Buffer) {
+    const tmpFile = getTmpFilePathSync(filepath)
+    processFile(filepath, 'r', (readFileDescriptor) => {
+        processFile(tmpFile, 'w', (writeFileDescriptor) => {
+            fs.writeSync(writeFileDescriptor, id3Tag)
+            streamOriginalIntoNewFileSync(readFileDescriptor, writeFileDescriptor)
+        })
+    })
+    fs.unlinkSync(filepath)
+    fs.renameSync(tmpFile, filepath)
+}
+
+export function writeId3TagToFileAsync(filepath: string, id3Tag: Buffer, callback: (err: Error|null) => void) {
+    getTmpFileAsync(filepath, (err, tmpFile) => {
+        if(err || !tmpFile) {
+            return callback(err)
+        }
+
+        processFileAsync(filepath, 'r', async (readFileDescriptor) => {
+            return processFileAsync(tmpFile, 'w', async (writeFileDescriptor) => {
+                await fsWritePromise(writeFileDescriptor, id3Tag)
+                await streamOriginalIntoNewFileAsync(readFileDescriptor, writeFileDescriptor)
+            })
+        }).then(async () => {
+            await fsUnlinkPromise(filepath)
+            await fsRenamePromise(tmpFile, filepath)
+            callback(null)
+        }).catch((error) => {
+            callback(error)
+        })
+    })
+}
+
+function getTmpFilePathSync(filepath: string): string {
+    const parsedPath = path.parse(filepath)
+    return tmp.tmpNameSync({
+        tmpdir: parsedPath.dir,
+        template: `${parsedPath.base}.tmp-XXXXXX`,
+    })
+}
+
+function getTmpFileAsync(filepath: string, callback: tmp.TmpNameCallback) {
+    const parsedPath = path.parse(filepath)
+    tmp.tmpName({
+        tmpdir: parsedPath.dir,
+        template: `${parsedPath.base}.tmp-XXXXXX`,
+    }, (err, filename) => {
+        callback(err, filename)
+    })
+}
+
+function streamOriginalIntoNewFileSync(readFileDescriptor: number, writeFileDescriptor: number) {
+    const buffer = Buffer.alloc(FileBufferSize)
+    let data
+    while((data = getNextBufferSubarraySync(readFileDescriptor, buffer)).length) {
+        const id3TagPosition = findId3TagPosition(data)
```

Comment:
There is a bug, we may miss the tag if across two readings, we need to use the same trick than in read.